<a href="https://colab.research.google.com/github/EJmpa/Hamoye_Stage_B/blob/main/Hamoye_Stage_B_ASSESMENT.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Library imporation

In [1]:
# Library imporation

import pandas as pd
import seaborn as sns
import numpy as np
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import MinMaxScaler
from sklearn.metrics import mean_absolute_error

In [2]:
from google.colab import files

uploaded = files.upload()

Saving energydata_complete.csv to energydata_complete.csv


# Read csv file into DataFrame

In [3]:
df = pd.read_csv("energydata_complete.csv")

df.set_index('date', inplace=True)
print(df.shape)
df.head()

(19735, 28)


Appliances  lights     T1       RH_1    T2       RH_2  \
date                                                                         
2016-01-11 17:00:00          60      30  19.89  47.596667  19.2  44.790000   
2016-01-11 17:10:00          60      30  19.89  46.693333  19.2  44.722500   
2016-01-11 17:20:00          50      30  19.89  46.300000  19.2  44.626667   
2016-01-11 17:30:00          50      40  19.89  46.066667  19.2  44.590000   
2016-01-11 17:40:00          60      40  19.89  46.333333  19.2  44.530000   

                        T3       RH_3         T4       RH_4  ...         T9  \
date                                                         ...              
2016-01-11 17:00:00  19.79  44.730000  19.000000  45.566667  ...  17.033333   
2016-01-11 17:10:00  19.79  44.790000  19.000000  45.992500  ...  17.066667   
2016-01-11 17:20:00  19.79  44.933333  18.926667  45.890000  ...  17.000000   
2016-01-11 17:30:00  19.79  45.000000  18.890000  45.723333  ...  17.000000   
2016-01-11 17:40:00  19.79  45.000000  18.890000  45.530000  ...  17.000000   

                      RH_9     T_out  Press_mm_hg  RH_out  Windspeed  \
date                                                                   
2016-01-11 17:00:00  45.53  6.600000        733.5    92.0   7.000000   
2016-01-11 17:10:00  45.56  6.483333        733.6    92.0   6.666667   
2016-01-11 17:20:00  45.50  6.366667        733.7    92.0   6.333333   
2016-01-11 17:30:00  45.40  6.250000        733.8    92.0   6.000000   
2016-01-11 17:40:00  45.40  6.133333        733.9    92.0   5.666667   

                     Visibility  Tdewpoint        rv1        rv2  
date                                                              
2016-01-11 17:00:00   63.000000        5.3  13.275433  13.275433  
2016-01-11 17:10:00   59.166667        5.2  18.606195  18.606195  
2016-01-11 17:20:00   55.333333        5.1  28.642668  28.642668  
2016-01-11 17:30:00   51.500000        5.0  45.410389  45.410389  
2016-01-11 17:40:00   47.666667        4.9  10.084097  10.084097  

[5 rows x 28 columns]

# Question 17

**From the dataset, fit a linear model on the relationship between the temperature in the living room in Celsius (x = T2) and the temperature outside the building (y = T6). What is the Root Mean Squared error in three D.P?**

In [4]:
features = df[['T2']]
target1 = df['T6']

model = LinearRegression()

#fit the model to the training dataset
model.fit(features, target1)


#obtain predictions
y_hat = model.predict(features)

In [5]:
#  Root Mean Square Error (RMSE)

from sklearn.metrics import mean_squared_error
rmse = np.sqrt(mean_squared_error(target1, y_hat))
print('\nRoot Mean Square Error:', round(rmse, 3))


Root Mean Square Error: 3.644


# Question 18

**Remove the following columns: `[“date”, “lights”]`. The target variable is “Appliances”. Use a 70-30 train-test set split with a  random state of 42 (for reproducibility). Normalize the dataset using the MinMaxScaler (Hint: Use the MinMaxScaler fit_transform and transform methods on the train and test set respectively). Run a multiple linear regression using the training set. Answer the following questions:**

**What is the Mean Absolute Error (in three decimal places) for the  training set?**

In [6]:
# Remove columns "date" and "lights"

# Define feature variables (X) and target variable (y)
X = df.drop(columns=["Appliances", "lights"])
y = df["Appliances"]

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# Normalize the dataset using MinMaxScaler
scaler = MinMaxScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

# Create and fit the Multiple Linear Regression model
model = LinearRegression()
model.fit(X_train, y_train)

# Make predictions on the training set
y_pred_train = model.predict(X_train)

# Calculate the Mean Absolute Error (MAE) for the training set
mae_train = mean_absolute_error(y_train, y_pred_train)

print(f'Mean Absolute Error (MAE) for the training set: {mae_train:.3f}')


Mean Absolute Error (MAE) for the training set: 53.742


# Question 19

**What is the Root Mean Squared Error (in three decimal places) for the training set?**

In [7]:
# Mean Squared Error (MSE) for the training set
mse_train = mean_squared_error(y_train, y_pred_train)

# Root Mean Squared Error (RMSE) for the training set
rmse_train = (mse_train ** 0.5)

print(f'Root Mean Squared Error (RMSE) for the training set: {rmse_train:.3f}')

Root Mean Squared Error (RMSE) for the training set: 95.216


# Question 20

**What is the Mean Absolute Error (in three decimal places) for test set?**

In [8]:
# Make predictions on the test set
y_pred_test = model.predict(X_test)

# Calculate the Mean Absolute Error (MAE) for the test set
mae_test = mean_absolute_error(y_test, y_pred_test)

print(f'Mean Absolute Error (MAE) for the test set: {mae_test:.3f}')

Mean Absolute Error (MAE) for the test set: 53.643


# Question 21

**What is the Root Mean Squared Error (in three decimal places) for test set?**

In [9]:
# Calculate the Mean Squared Error (MSE) for the test set
mse_test = mean_squared_error(y_test, y_pred_test)

# Calculate the Root Mean Squared Error (RMSE) for the test set
rmse_test = (mse_test ** 0.5)

print(f'Root Mean Squared Error (RMSE) for the test set: {rmse_test:.3f}')

Root Mean Squared Error (RMSE) for the test set: 93.640


# Question 22
**Did the Model above overfit to the training set?**

The model will be `Overfitting` if the `RMSE on the test set` is `much higher` than the `RMSE on the training set`. If the model's performance on both the training set and the test set is similar or if it performs better on the test set, it is less likely to be overfitting. This indicates that the model is likely to be a good fit for the data.

`RMSE on the test set`: 93.637\
`RMSE on the train set`: 95.216

Since the RMSE of the test and training set is similar and that of test set is lower than training set. The model is likely a good fit for the data

**Answer**: **NO**

# Question 23

**Train a ridge regression model with default parameters. Is there any change to the root mean squared error (RMSE) when evaluated on the test set?**

In [10]:
from sklearn.linear_model import Ridge
ridge_reg = Ridge(alpha=0.5)
ridge_reg.fit(X_train, y_train)

# Make predictions on the test set using the Ridge Regression model
y_pred_ridge = ridge_reg.predict(X_test)

# Calculate the Root Mean Squared Error (RMSE) for both models on the test set
rmse_linear = mean_squared_error(y_test, y_pred_test, squared=False)
rmse_ridge = mean_squared_error(y_test, y_pred_ridge, squared=False)

print(f'RMSE - Linear Regression Model: {rmse_linear:.3f}')
print(f'RMSE - Ridge Regression Model: {rmse_ridge:.3f}')

RMSE - Linear Regression Model: 93.640
RMSE - Ridge Regression Model: 93.667


In [11]:
print(rmse_linear - rmse_ridge)

-0.026681015256130536


**Answer: Yes there is a slight difference**

# Question 24

**Train a lasso regression model with default value and obtain the new feature weights with it. How many of the features have non-zero feature weights?**

In [12]:
# Create and fit the Lasso Regression model with default parameters
from sklearn.linear_model import Lasso
lasso_reg = Lasso(alpha=0.001)
lasso_reg.fit(X_train, y_train)

# Count the number of features with non-zero feature weights
non_zero_features = sum(lasso_reg.coef_ != 0)

print(f'Number of features with non-zero feature weights: {non_zero_features}')

Number of features with non-zero feature weights: 26


# Question 25

**What is the new RMSE with the Lasso Regression on the test set?**

In [13]:
# Make predictions on the test set using the Lasso Regression model
y_pred_lasso = lasso_reg.predict(X_test)

# Calculate the Mean Squared Error (MSE) for the test set
mse_test = mean_squared_error(y_test, y_pred_lasso)

# Calculate the Root Mean Squared Error (RMSE) for the test set
rmse_test = np.sqrt(mse_test)

print(f'Root Mean Squared Error (RMSE) with Lasso Regression on the test set: {rmse_test:.3f}')

Root Mean Squared Error (RMSE) with Lasso Regression on the test set: 93.641
